### PR TITLE
Add `all_triangles` generator yielding all unique triangles in a graph

### PR DESCRIPTION
**Pull request by georakom**
Implemented an `all_triangles` function that efficiently yields all unique triangles in an undirected graph.
The algorithm follows a "Forward" principle by enforcing a node ordering to avoid redundant comparisons and double-counting. For each node u, it examines neighbors v > u and yields triangles formed by the intersection of u's and v's neighbors that are greater than v. 
The function optionally supports a nodes parameter, restricting enumeration to triangles where all three nodes are in the specified set.

This is my first pull request here and I would gladly appreciate any comments and improvements!

### CODE REVIEW

**Comment by georakom** [comment]
Thanks again for the reponse @rossbar ! 
I took a closer look at nx.simple_cycles, and while it definitely works in practice, it's not an ideal tool for triangle enumeration in practice. The function is built for general purpose cycle enumeration, which can bring in quite a bit of overhead, especially for large or undirected graphs. Also, it explores components and paths far beyond what's needed for triangle detection. In my use case, I found that a dedicated triangle generator is an order of magnitude faster, and more memory efficient.
I'd be happy to modify the PR to implement a unique_triangles(G) generator instead of a counting function, if that would better align with NetworkX’s design.

**Comment by rossbar** [comment]
>  I'd be happy to modify the PR to implement a unique_triangles(G) generator instead of a counting function, if that would better align with NetworkX’s design.

It would be more compelling to me at least! Re: naming, I'd propose `all_triangles` which matches a naming pattern used elsewhere, see e.g. `all_triads`. In fact, that's another bit of existing functionality that meets this need: `nx.all_triads(G.to_directed())`.

File: networkx/algorithms/cluster.py [hunk 1]
```diff
@@ -247,6 +247,72 @@ def wt(u, v):
...
+    --------
+    >>> G = nx.complete_graph(4)
+    >>> list(all_triangles(G))
+    [(0, 1, 2), (0, 1, 3), (0, 2, 3), (1, 2, 3)]
+    """
+    # If nodes is None, then yield all the triangles of the graph
+    if nodes is None:
+        # Pre-computation based on fixed ID ordering
+        neighbor_sets = {
+            node: {nbr for nbr in G.neighbors(node) if nbr > node} for node in G
```
Comment (rossbar) [inline networkx/algorithms/cluster.py]: This optimization is invalid in networkx - it is never guaranteed that nodes are ordered this way. In fact, it's not even guaranteed that nodes are comparable!

**Comment by georakom** [comment]
Thank you for the feedback @rossbar ! I’ve gone ahead and implemented the generator version as all_triangles, following the naming pattern you mentioned.

**Review by rossbar — COMMENTED** [review]
Make sure to remove the extraneous changes to the other files that have nothing to do with this PR, which would be required before this gets any in-depth review.

Also - you will likely garner more traction from the community if you can provide stronger motivation for the addition. For example, what problem were you trying to solve where this functionality was necessary? Similarly, why not just use one of the other proposed methods - do we really need a new function for this?

**Comment by georakom** [comment]
Thanks again for the feedback @rossbar . I have now removed all unrelated file changes from the PR to keep it clean. I have also addressed the issue regarding node comparability. Specifically, the implementation now uses _str(node)_-based ordering to ensure consistent and safe behavior regardless of node type, which avoids assumptions about comparability or orderability.
As for motivation, I am currently doing a thesis on triangle counting in undirected graphs (in Python), and during my research, I explored various enumeration techniques. This led me to try implementing one directly in NetworkX, adapted to follow the library’s design principles. 
I'm happy to further adjust the code or design if needed.

**Comment by georakom** [comment]
Thank you @rossbar, I implemented now a generator that doesn't rely on any node ordering. Now the implementation yields the unique triangles by using purely the adjacency structure and set intersections. However, the current implementation might suffer a bit for larger graphs, and I will work on refining it to incorperate some form of visited ordering to avoid redundant checks.

**Comment by amcandio** [comment]
This implementation is storing all triangles, which means memory consumption is cubical on the amount of nodes in the worst case. I know in NetworkX you cannot depend on ordering the nodes but you can do a linear memory solution.

You assign an incremental id to each node as you see them for the first time. Then you only yield the ones with form (u, v, w) where id(u) < id(v) < id(w).

**Comment by georakom** [comment]
You are right, the implementation did need an improvement. I was thinking of some sort of "visited" ordering, but your idea is simpler to execute, and so I implemented it. The function now assigns an ID to the nodes needed and then computes intersections while avoiding redundant checks. Thank you for the suggestion, @amcandio!

File: networkx/algorithms/cluster.py [hunk 2]
```diff
@@ -247,6 +247,80 @@ def wt(u, v):
...
+        guaranteed to be unique.
+
+    Examples
+    --------
+    >>> G = nx.complete_graph(4)
+    >>> list(all_triangles(G))
+    [(0, 1, 2), (0, 1, 3), (0, 2, 3), (1, 2, 3)]
+    """
+    if nodes is None:
+        node_to_id = {node: i for i, node in enumerate(G.nodes())}
```
Comment (amcandio) [inline networkx/algorithms/cluster.py]: You can just iterate G

```suggestion
        node_to_id = {node: i for i, node in enumerate(G)}
```

File: networkx/algorithms/cluster.py [hunk 3]
```diff
@@ -247,6 +247,80 @@ def wt(u, v):
...
+
+        neighbor_sets = {
+            node: {nbr for nbr in G[node] if node_to_id[nbr] > node_to_id[node]}
+            for node in G.nodes()
+        }
+        for u in G.nodes():
+            for v in neighbor_sets[u]:
+                common = neighbor_sets[u] & neighbor_sets[v]
+                for w in common:
+                    yield u, v, w
```
Comment (amcandio) [inline networkx/algorithms/cluster.py]: This is consuming n^2 extra memory. You don't need to cache the nodes, for you could do:

```suggestion
        for u in G:
            for v in G[u]:
                if node_to_id[v] <= node_to_id[u]:
                    continue
                for w in G[u] & G[v]:
                    if node_to_id[w] <= node_to_id[v]:
                        continue
                    yield (u, v, w)
```

Comment (georakom) [reply networkx/algorithms/cluster.py]: Yeah you are right, but for a large graph will they have similar performance? Also " G[u] & G[v] " is not supported for Atlas view

Comment (amcandio) [reply networkx/algorithms/cluster.py]: Yes, it could even be faster due to less memory allocations. My bad on the &, we will have to use neighbors method

Comment (georakom) [reply networkx/algorithms/cluster.py]: But from my understanding, NetworkX has G[nodes] that we can't operate intersections with the neighbors of nodes, so I think building the neighbor_sets is needed either way. Edit: We can only intersect if we convert the node's neighbors into a set()

Comment (georakom) [reply networkx/algorithms/cluster.py]: Thank you for the suggestion, I went on and applied it!

File: networkx/algorithms/cluster.py [hunk 4]
```diff
@@ -254,6 +254,70 @@ def wt(u, v):
...
+        # Keep only relevant nodes for counting triangles (nodes and their neighbors)
+        relevant_nodes = nodes_set | {
+            nbr for node in nodes_set for nbr in G.neighbors(node)
+        }
+
+        # Need to assign IDs to all neighbors of relevant_nodes as well
+        all_needed = relevant_nodes | {
+            nbr for node in relevant_nodes for nbr in G.neighbors(node)
+        }
+
```
Comment (amcandio) [inline networkx/algorithms/cluster.py]: We can dedupe this by only setting an id for relevant nodes. Then you access the id by doing `node_to_id.get(u, -1)`. This way you can use the same routine for both

Comment (georakom) [reply networkx/algorithms/cluster.py]: Thanks again @amcandio ! You are right, the `all_needed` was a too defensive way to avoid key errors, but this fixes that issue. I went ahead and changed this as well!

File: networkx/algorithms/cluster.py [hunk 5]
```diff
@@ -254,6 +254,65 @@ def wt(u, v):
...
+        for u in G:
+            for v in G[u]:
+                if node_to_id[v] <= node_to_id[u]:
+                    continue
+                for w in set(G[u]) & set(G[v]):
+                    if node_to_id[w] <= node_to_id[v]:
+                        continue
+                    yield u, v, w
+    else:
+        nodes_set = {nodes} if nodes in G else set(nodes)
```
Comment (amcandio) [inline networkx/algorithms/cluster.py]: This can backfire if you pass a node that is not in G. Why not just receive an iterable? Users can pass a single element set

File: networkx/algorithms/cluster.py [hunk 6]
```diff
@@ -254,6 +254,65 @@ def wt(u, v):
...
+
+        for u in relevant_nodes:
+            for v in G[u]:
+                if node_to_id.get(v, -1) <= node_to_id[u]:
+                    continue
+                for w in set(G[u]) & set(G[v]):
+                    if node_to_id.get(w, -1) <= node_to_id[v]:
+                        continue
+                    if u in nodes_set or v in nodes_set or w in nodes_set:
+                        yield u, v, w
```
Comment (amcandio) [inline networkx/algorithms/cluster.py]: What was missing from my previous comment was that now you can dedupe the for loops. No need to have both

Comment (amcandio) [reply networkx/algorithms/cluster.py]: By both I mean, whether you pass "nodes" or not

Comment (georakom) [reply networkx/algorithms/cluster.py]: You mean something like this?
```
for u in iterator:
        for v in G[u]:
            if node_to_id.get(v, -1) <= node_to_id[u]:
                continue
            for w in set(G[u]) & set(G[v]):
                if node_to_id.get(w, -1) <= node_to_id[v]:
                    continue
                if nodes_set is None or u in nodes_set or v in nodes_set or w in nodes_set:
                    yield u, v, w
```
I thought of this as well, but I just thought that it might cause a slight overhead if nodes is None, because we do redundant checkings.

Comment (dschult) [reply networkx/algorithms/cluster.py]: I think what @amcandio meant was you don't need to check that the nodes are in nodes_set anymore because you have bypassed them if they aren't in nodes_set.

Also, this might be a place where `G.nbunch_iter(nodes)` is useful. When the input is a collection of nodes, we typically call it an "nbunch" and feed it to `G.nbunch_iter` which handles `None` and ignores nodes not in `G`. 
`nodes_set = set(G.nbunch_iter(nodes))`

You can also avoid lots of set creation using `w not in unbrs` instead of `&`. It's worth seeing if that's better.
And I added the minor optimization of using `G._adj[u]` instead of `G[u]` which avoids the `Adjacency/Atlas` wrappers -- useful when you know you are only reading the adjacency dict.
```python
for u in nodes_set:
        u_id = node_to_id[u]
        unbrs = set(G._adj[u])
        for v in unbrs:
            v_id = node_to_id.get(v, -1)
            if v_id <= u_id:
                continue
            for w in G._adj[v]:
                if w not in unbrs or (node_to_id.get(w, -1) <= v_id):
                    continue
                yield u, v, w
```
The last `if` is no longer needed because we already checked if v and w are in nodes_set.  It means the whole function is shorter too. :)

Comment (amcandio) [reply networkx/algorithms/cluster.py]: Sorry, was very unclear. I meant that you don't need to duplicate the 3 nested loops. You can make the second case generalize the first one

Comment (georakom) [reply networkx/algorithms/cluster.py]: Thank you for all the suggestions! I went ahead and changed everything that had to be changed. Also, using `w not in u_nbrs` had a very similar performance for medium and small graphs, but for larger graphs `&` seemed to perform a bit better, so I used the intersection method. Everything else was changed based on your suggestions :)

File: networkx/algorithms/cluster.py [hunk 7]
```diff
@@ -254,6 +254,53 @@ def wt(u, v):
...
+    """
+    nodes_set = set(G.nbunch_iter(nodes))
+    relevant_nodes = nodes_set | {
+        nbr for node in nodes_set for nbr in G.neighbors(node)
+    }
+    node_to_id = {node: i for i, node in enumerate(relevant_nodes)}
+
+    for u in nodes_set:
+        u_id = node_to_id[u]
+        u_nbrs = set(G._adj[u])
```
Comment (amcandio) [inline networkx/algorithms/cluster.py]: Same here (it also allows us to avoid creating a new set)

```suggestion
        u_nbrs = G._adj[u].keys()
```

File: networkx/algorithms/cluster.py [hunk 8]
```diff
@@ -254,6 +254,53 @@ def wt(u, v):
...
+        for v in u_nbrs:
+            v_id = node_to_id.get(v, -1)
+            if v_id <= u_id:
+                continue
+            v_nbrs = set(G._adj[v])
```
Comment (amcandio) [inline networkx/algorithms/cluster.py]: Same here:
```suggestion
            v_nbrs = G._adj[v].keys()
```

File: networkx/algorithms/cluster.py [hunk 9]
```diff
@@ -254,6 +254,53 @@ def wt(u, v):
...
+    Examples
+    --------
+    >>> G = nx.complete_graph(4)
+    >>> list(all_triangles(G))
+    [(0, 1, 2), (0, 1, 3), (0, 2, 3), (1, 2, 3)]
+    """
+    nodes_set = set(G.nbunch_iter(nodes))
+    relevant_nodes = nodes_set | {
+        nbr for node in nodes_set for nbr in G.neighbors(node)
+    }
```
Comment (amcandio) [inline networkx/algorithms/cluster.py]: This can make output inconsistent over different runs, which is a nice property to have (predictable output, easier to write unit tests for users). I already check and `chain` is already imported

```suggestion
    nodes_set = dict.fromkeys(G.nbunch_iter(nodes))
    relevant_nodes = chain(
        nodes_set,
        (nbr for node in node_set for nbr in G.neighbors(node) if nbr not in nodes_set)
    )
```

File: networkx/algorithms/cluster.py [hunk 10]
```diff
@@ -254,6 +254,53 @@ def wt(u, v):
         yield (i, dtotal, dbidirectional, float(directed_triangles))
 
 
+@not_implemented_for("directed")
+@nx._dispatchable
+def all_triangles(G, nodes=None):
+    """
+    Yields all unique triangles in an undirected graph using ID assignments and set intersections.
```
Comment (amcandio) [inline networkx/algorithms/cluster.py]: We should probably hide implementation details here. I think they might confuse users
```suggestion
    Yields all unique triangles in an undirected graph. A triangle is a set of three vertices in a graph where 
    each pair of vertices is connected by an edge.
```

File: networkx/algorithms/cluster.py [hunk 11]
```diff
@@ -254,6 +254,53 @@ def wt(u, v):
...
+
+    nodes : node, iterable of nodes, or None (default=None)
+        If a node or iterable of nodes is given, only triangles involving those nodes are yielded.
+        If None, yields all unique triangles in the graph.
+
+    Yields
+    -------
+    tuple
+        A tuple of three nodes forming a triangle (u, v, w), ordered based on assigned ID,
+        guaranteed to be unique.
```
Comment (amcandio) [inline networkx/algorithms/cluster.py]: Same here. What do you think?

```suggestion
        A tuple of three nodes forming a triangle (u, v, w), guaranteed to be unique.
```

File: networkx/algorithms/tests/test_cluster.py [hunk 12]
```diff
@@ -77,6 +79,71 @@ def test_k5(self):
         assert nx.triangles(G, 3) == 5
```
Comment (amcandio) [inline networkx/algorithms/tests/test_cluster.py:79]: Can we add a unit test on a predictable graph? Like a complete graph

Comment (georakom) [reply networkx/algorithms/tests/test_cluster.py:79]: There is one a unit test already on a complete graph.

File: networkx/algorithms/cluster.py (hunk 11 above)
Comment (georakom) [reply networkx/algorithms/cluster.py]: I agree. You are right that the comments can be a bit more brief without implementation details, so I changed them!

**Comment by georakom** [comment]
Thank you for the suggestions @amcandio ! I will now have more flexibility for the unit tests and I will change them too now!

File: networkx/algorithms/tests/test_cluster.py (hunk 12 above)
Comment (amcandio) [reply networkx/algorithms/tests/test_cluster.py:79]: I meant that you can parameterize n on a complete graph. And then on the test case you infer the result by generating all combinations

**Review by amcandio — APPROVED** [review]
LGTM, great work!

File: networkx/algorithms/cluster.py [hunk 13]
```diff
@@ -254,6 +254,59 @@ def wt(u, v):
...
+    Examples
+    --------
+    >>> G = nx.complete_graph(4)
+    >>> list(all_triangles(G))
+    [(0, 1, 2), (0, 1, 3), (0, 2, 3), (1, 2, 3)]
+    """
+    nodes_set = dict.fromkeys(G.nbunch_iter(nodes))
+    relevant_nodes = chain(
+        nodes_set,
+        (
```
Comment (amcandio) [inline networkx/algorithms/cluster.py]: One optimization we could add is, if no nodes are passed, no need to consume this extra iteration. This would reduce running time for the default case, bc it removes the extra iteration

**Comment by amcandio** [comment]
Btw, let's update the description before shipping this

File: networkx/algorithms/cluster.py (hunk 13 above)
Comment (georakom) [reply networkx/algorithms/cluster.py]: Yes this was my oversight, but now I will change it! Thank you.

**Comment by georakom** [comment]
@amcandio You mean we should make the description more in detail? Or sharpen it?

**Comment by amcandio** [comment]
I think the description is for a previous approach?

**Comment by georakom** [comment]
@amcandio it was yes but I checked it and I thought it could fit to the new changes as well. However, I have changed it and I hope it's better now :)

**Comment by amcandio** [comment]
I'm confused. I don't see any triangle count method?

**Comment by georakom** [comment]
@amcandio oh excuse me I thought you were referring to the doc string in the function. Yes I will change the description of the PR. Thank you!

File: networkx/algorithms/cluster.py [hunk 14]
```diff
@@ -254,6 +254,63 @@ def wt(u, v):
...
+    tuple
+        A tuple of three nodes forming a triangle (u, v, w), with u < v < w based on internal ordering.
+
+    Examples
+    --------
+    >>> G = nx.complete_graph(4)
+    >>> list(all_triangles(G))
+    [(0, 1, 2), (0, 1, 3), (0, 2, 3), (1, 2, 3)]
+    """
+    nodes_set = dict.fromkeys(G.nbunch_iter(nodes))
```
Comment (amcandio) [inline networkx/algorithms/cluster.py]: Last optimization: you don't need to create this new dict if nodes is None. You can use G for that

Comment (georakom) [reply networkx/algorithms/cluster.py]: But we have to iterate in `nodes_set` either way since we have every option to operate in the same for loop. You mean I should put `nodes_set = set(G)` instead of the dict if nodes is None?

Comment (dschult) [reply networkx/algorithms/cluster.py]: I think the suggestion is to have `node_set = G`.  Which also reminds us that `node_set` is not a set anymore and should probably be renamed.  How about something like:
```python
if nodes is None:
    nodes = relevant_nodes = G
else:
    nodes = dict.fromkeys(G.nbunch_iter(nodes))
    relevant_nodes = chain(...
    )
node_to_id = {node: i for i, node in enumerate(relevant_nodes)}

for u in nodes:
    ...
```
And we should probably rename the input parameter from `nodes` to `nbunch` to make it clear that we are using the standard networkx handling of a collection of nodes (nbunch is described in the NetworkX glossary).

File: networkx/algorithms/cluster.py [hunk 15]
```diff
@@ -254,6 +254,63 @@ def wt(u, v):
         yield (i, dtotal, dbidirectional, float(directed_triangles))
 
 
+@not_implemented_for("directed")
+@nx._dispatchable
+def all_triangles(G, nodes=None):
+    """
+    Yields all unique triangles in an undirected graph using a node ordering strategy.
```
Comment (Peiffap) [inline networkx/algorithms/cluster.py]: The implementation details should be in a `Notes` section imo, where it can be used to justify an asymptotic run time complexity.
```suggestion
    Yields all unique triangles in an undirected graph.
```

File: networkx/algorithms/cluster.py [hunk 16]
```diff
@@ -254,6 +254,63 @@ def wt(u, v):
...
+    A triangle is a set of three distinct nodes where each node is connected to the other two.
+
+    Parameters
+    ----------
+    G : NetworkX graph
+        An undirected graph. Self-loops and multiple edges are ignored.
+
+    nodes : node, iterable of nodes, or None (default=None)
+        If a node or iterable of nodes is given, only triangles involving those nodes are yielded.
+        If None, yields all unique triangles in the graph.
```
Comment (Peiffap) [inline networkx/algorithms/cluster.py]:
```suggestion
        If ``None``, yields all unique triangles in the graph.
```

File: networkx/algorithms/cluster.py [hunk 17]
```diff
@@ -254,6 +254,63 @@ def wt(u, v):
...
+        u_nbrs = G._adj[u].keys()
+        for v in u_nbrs:
+            v_id = node_to_id.get(v, -1)
+            if v_id <= u_id:
+                continue
+            v_nbrs = G._adj[v].keys()
+            for w in v_nbrs & u_nbrs:
+                if node_to_id.get(w, -1) <= v_id:
+                    continue
+                yield u, v, w
```
Comment (Peiffap) [inline networkx/algorithms/cluster.py]:
```suggestion
                if node_to_id.get(w, -1) > v_id:
                    yield u, v, w
```

File: networkx/algorithms/cluster.py [hunk 18]
```diff
@@ -254,6 +254,63 @@ def wt(u, v):
...
+    Yields
+    -------
+    tuple
+        A tuple of three nodes forming a triangle (u, v, w), with u < v < w based on internal ordering.
+
+    Examples
+    --------
+    >>> G = nx.complete_graph(4)
+    >>> list(all_triangles(G))
+    [(0, 1, 2), (0, 1, 3), (0, 2, 3), (1, 2, 3)]
```
Comment (Peiffap) [inline networkx/algorithms/cluster.py]: Should this (and the tests which use lists too) use sets instead? There isn't any guarantee in the function description about the order in which tuples are returned, so if the implementation changes at any point, it'd be nice to not have that be a breaking change.

File: networkx/algorithms/cluster.py (hunk 14 above)
Comment (georakom) [reply networkx/algorithms/cluster.py]: Thank you. Fixed it now.

File: networkx/algorithms/cluster.py [hunk 19]
```diff
@@ -254,6 +254,63 @@ def wt(u, v):
...
+        An undirected graph. Self-loops and multiple edges are ignored.
+
+    nodes : node, iterable of nodes, or None (default=None)
+        If a node or iterable of nodes is given, only triangles involving those nodes are yielded.
+        If None, yields all unique triangles in the graph.
+
+    Yields
+    -------
+    tuple
+        A tuple of three nodes forming a triangle (u, v, w), with u < v < w based on internal ordering.
```
Comment (Peiffap) [inline networkx/algorithms/cluster.py]: Iiuc, someone calling the function won't know anything about this ordering (nor is it certain to stay the same in the future), so I'm not sure it's good to specify.
```suggestion
        A tuple of three nodes forming a triangle ``(u, v, w)``.
```

File: networkx/algorithms/cluster.py [hunk 20]
```diff
@@ -254,6 +254,63 @@ def wt(u, v):
...
+
+    A triangle is a set of three distinct nodes where each node is connected to the other two.
+
+    Parameters
+    ----------
+    G : NetworkX graph
+        An undirected graph. Self-loops and multiple edges are ignored.
+
+    nodes : node, iterable of nodes, or None (default=None)
+        If a node or iterable of nodes is given, only triangles involving those nodes are yielded.
```
Comment (Peiffap) [inline networkx/algorithms/cluster.py]: Wording nit: it's unclear from this sentence whether only triangles that have all of their nodes in `nodes` are considered, or only triangles that have at least one of their nodes in `nodes` (or only triangles that have all of the nodes in `nodes`—although that would be an unlikely interpretation...).
```suggestion
        If a node or iterable of nodes is given, only triangles with all their nodes in `nodes` are yielded.
```

File: networkx/algorithms/cluster.py [hunk 21]
```diff
@@ -254,6 +254,63 @@ def wt(u, v):
...
+def all_triangles(G, nodes=None):
+    """
+    Yields all unique triangles in an undirected graph using a node ordering strategy.
+
+    A triangle is a set of three distinct nodes where each node is connected to the other two.
+
+    Parameters
+    ----------
+    G : NetworkX graph
+        An undirected graph. Self-loops and multiple edges are ignored.
```
Comment (Peiffap) [inline networkx/algorithms/cluster.py]: Minor nit: I believe this doesn't matter (self-loops and multiedges shouldn't affect the result), and it might just end up confusing users ("will this function work for multigraphs?").
```suggestion
        An undirected graph.
```

File: networkx/algorithms/tests/test_cluster.py [hunk 22]
```diff
@@ -2,6 +2,8 @@
 
 import networkx as nx
 
+from ..cluster import all_triangles
+
```
Comment (Peiffap) [inline networkx/algorithms/tests/test_cluster.py]: I think this won't be necessary anymore once the new function is properly exposed in the top-level namespace. You'll be able to just call `nx.all_triangles` then.

File: networkx/algorithms/tests/test_cluster.py [hunk 23]
```diff
@@ -77,6 +79,71 @@ def test_k5(self):
...
+        ]
+    )
+    triangles = list(all_triangles(G, nodes=[0, 1]))
+    assert triangles == [(0, 1, 2)]
+
+
+def test_all_triangles_no_duplicates():
+    G = nx.complete_graph(4)  # 4 nodes, 4-choose-3 = 4 triangles
+    triangles = list(all_triangles(G))
+    assert len(triangles) == 4
```
Comment (Peiffap) [inline networkx/algorithms/tests/test_cluster.py]: This is redundant if `test_all_triangles_complete_graph_exact` is kept.

File: networkx/algorithms/tests/test_cluster.py [hunk 24]
```diff
@@ -77,6 +79,71 @@ def test_k5(self):
...
+    expected = [
+        (0, 1, 2),
+        (0, 1, 3),
+        (0, 2, 3),
+        (1, 2, 3),
+    ]
+
+    assert triangles == expected
+
+
```
Comment (Peiffap) [inline networkx/algorithms/tests/test_cluster.py:173]: Can we add a test where the graph has no triangles too ? And can we add a test where the graph has triangles but none of them in `nodes`?

Some more tests we should consider adding (but that I don't feel are strictly necessary):

- Test that this raises `NetworkXNotImplemented` for directed graphs
- Parametrize the existing tests for `nx.Graph` and `nx.MultiGraph`.

**Review by Peiffap — COMMENTED** [review]
This is nice! Some comments below, but feel free to push back if you disagree :).

Can this be used to speed up `nx.simple_cycles` for the undirected, `length_bound = 3` case? It would be nice to benchmark that... It might also be interesting to add it to the See Also section for that function in that case. This should probably all be done in a separate PR, though, if we decide it's worthwhile.

**Comment by Peiffap** [comment]
I was reviewing this before you made the latest changes—let's make sure my suggestions don't undo some of the other changes you've made!

File: networkx/algorithms/tests/test_cluster.py (hunk 24 above)
Comment (georakom) [reply networkx/algorithms/tests/test_cluster.py:173]: Yes, I will add these right now!

File: networkx/algorithms/cluster.py (hunk 18 above)
Comment (georakom) [reply networkx/algorithms/cluster.py]: I completely agree. If I am not mistaken, the `node_to_id` ordering is based on how the graph is inserted, right? If I build manually a graph and I do `G.add_nodes_from([2, 0, 5])`, I think the triangle (0,2,5) will be yielded like (2,0,5). I am not 100% sure so let me know!

Comment (Peiffap) [reply networkx/algorithms/cluster.py]: That's actually a separate point I believe—there are two levels at which ordering can matter:

1. This note is about the order in which the tuples are returned, e.g. `all_triangles(G)` could yield `(0, 1, 2)` then `(0, 1, 3)` etc., but yielding them in a different order would still be correct—making the examples and tests use lists means this order is somehow "checked", when I don't believe it should be.
2. The order of nodes inside a tuple is also relevant, and another comment I made addresses that question. A tuple `(u, v, w)` can also be given as `(v, w, u)`, etc. Since technically for an undirected triangle, all 6 node orderings are valid (unlike for a cycle of length 4, where the order matters), this could also be a set, but I think by analogy with `nx.simple_cycles`, an argument can be made that it's perhaps more consistent to return a tuple. I'd be curious to hear other opinions on this.

Comment (georakom) [reply networkx/algorithms/cluster.py]: The tests and the example have a specific ordering on how the graph is built, and yes I agree that this can give the impression that the triangles are always yielded in a specific order that the human eye sees, which might not be the case. So I think I do need to remove the mention of `u < v < w` from the docstring and make the tests order-independent (using sets or frozensets).

File: networkx/algorithms/cluster.py (hunk 20 above)
Comment (georakom) [reply networkx/algorithms/cluster.py]: I wasn't very specific with it, I will describe it more concisely now, thank you! However your suggestion doesn't completely apply to what the `nbunch` is actually for. If an iterable of nodes is given, the triangles that are actually yielded are the ones that contain **atleast one** of the nodes from `nbunch`.

File: networkx/algorithms/cluster.py (hunk 18 above)
Comment (georakom) [reply networkx/algorithms/cluster.py]: What I could also do is to apply tests where the graphs are built with unorthodox ordering, and use there sets.

File: networkx/algorithms/cluster.py (hunk 20 above)
Comment (Peiffap) [reply networkx/algorithms/cluster.py]: Welp... That illustrates my point I suppose!

File: networkx/algorithms/cluster.py (hunk 18 above)
Comment (Peiffap) [reply networkx/algorithms/cluster.py]: Maybe returning a tuple `(u, v, w)` but running the tests with set (something like `assert {set(t) for t in nx.all_triangles(G)} == {{u1, v1, w1}, {u2, v2, w2}, ...}` makes the most sense then!

Comment (georakom) [reply networkx/algorithms/cluster.py]: Okay perfect, will apply it now as well then!

**Comment by georakom** [comment]
Thank you for all the suggestions @Peiffap ! I went ahead and applied every single one of them.
- Changed the comments and put notes for implementation description and complexity
- Applied set-based comparisons in the tests
- Added the tests you recommended!
I am not very experienced so I added every single suggestion you gave manually just in case it doesn't undo other changes, I hope that's fine :) !

File: networkx/algorithms/cluster.py [hunk 25]
```diff
@@ -254,6 +254,63 @@ def wt(u, v):
...
+
+    A triangle is a set of three distinct nodes where each node is connected to the other two.
+
+    Parameters
+    ----------
+    G : NetworkX graph
+        An undirected graph.
+
+    nbunch : node, iterable of nodes, or None (default=None)
+        If a node or iterable of nodes, only triangles involving at least one node in nbunch are yielded.
```
Comment (Peiffap) [inline networkx/algorithms/cluster.py]:
```suggestion
        If a node or iterable of nodes, only triangles involving at least one node in `nbunch` are yielded.
```

File: networkx/algorithms/cluster.py [hunk 26]
```diff
@@ -254,6 +254,63 @@ def wt(u, v):
...
+        If ``None``, yields all unique triangles in the graph.
+
+    Yields
+    -------
+    tuple
+         A tuple of three nodes forming a triangle ``(u, v, w)``.
```
Comment (Peiffap) [inline networkx/algorithms/cluster.py]:
```suggestion
        A tuple of three nodes forming a triangle ``(u, v, w)``.
```

File: networkx/algorithms/cluster.py [hunk 27]
```diff
@@ -254,6 +254,63 @@ def wt(u, v):
...
+    Examples
+    --------
+    >>> G = nx.complete_graph(4)
+    >>> sorted([sorted(t) for t in all_triangles(G)])
+    [[0, 1, 2], [0, 1, 3], [0, 2, 3], [1, 2, 3]]
+
+    Notes
+    -----
+    This algorithm ensures each triangle is yielded once using an internal node ordering.
+    It runs in O(m * d) time in the worst case, where m is the number of edges and d is the maximum degree.
```
Comment (Peiffap) [inline networkx/algorithms/cluster.py]: Where did you get this complexity? Shouldn't it be `O(m * d**2)` because you compute a set intersection in the loop on L309?
```suggestion
    It runs in ``O(m * d)`` time in the worst case, where ``m`` is the number of edges and ``d`` is the maximum degree.
```

File: networkx/algorithms/cluster.py (hunk 26 above)
Comment (georakom) [reply networkx/algorithms/cluster.py]: I need to be more careful with such details, thank you!

File: networkx/algorithms/cluster.py (hunk 27 above)
Comment (georakom) [reply networkx/algorithms/cluster.py]: A set intersection of two sets of size at most `d` costs `O(d)` and not `O(d**2)`. I am not doing an inner nested scan over both neighbor lists since python’s set intersection does the efficient hashed lookup approach.

**Comment by georakom** [comment]
Thank you @Peiffap. I have now added the `all_triangles` function to `__all__` and also fixed the issues in the comments!

File: networkx/algorithms/tests/test_cluster.py [hunk 28]
```diff
@@ -77,6 +78,104 @@ def test_k5(self):
...
+    G = nx.Graph()
+    G.add_edges_from(
+        [
+            ("a", "b"),
+            ("b", "c"),
+            ("c", "a"),  # triangle: a-b-c
+        ]
+    )
+    triangles = list(nx.all_triangles(G))
+    assert set(map(frozenset, triangles)) == {frozenset({"a", "b", "c"})}
```
Comment (rossbar) [inline networkx/algorithms/tests/test_cluster.py]: Just a nit, but I prefer set comprehension:

```suggestion
    expected = {frozenset({"a", "b", "c"})}
    assert {frozenset(t) for t in nx.all_triangles)} == expected
```

File: networkx/algorithms/tests/test_cluster.py [hunk 29]
```diff
@@ -1,6 +1,7 @@
 import pytest
 
 import networkx as nx
+from networkx.exception import NetworkXNotImplemented
```
Comment (rossbar) [inline networkx/algorithms/tests/test_cluster.py]:
```suggestion
```

File: networkx/algorithms/tests/test_cluster.py [hunk 30]
```diff
@@ -77,6 +78,104 @@ def test_k5(self):
...
+        frozenset({1, 2, 3}),
+    }
+
+    assert set(map(frozenset, triangles)) == expected
+
+
+def test_all_triangles_directed_graph():
+    G = nx.DiGraph()
+    G.add_edges_from([(0, 1), (1, 2), (2, 0)])
+    with pytest.raises(NetworkXNotImplemented):
```
Comment (rossbar) [inline networkx/algorithms/tests/test_cluster.py]:
```suggestion
    with pytest.raises(nx.NetworkXNotImplemented):
```

File: networkx/algorithms/cluster.py (hunk 27 above)
Comment (Peiffap) [reply networkx/algorithms/cluster.py]: Ah so is it because we have the `continue` statement that we only end looping over the intersection of neighbors once? Because from my understanding, we are looping over all nodes, then all neighbors of that node, then all nodes in the intersection of the neighbors of both.

**Comment by georakom** [comment]
Thanks @rossbar . Yes the function's behavior on multigraphs is indeed the option 1, since I am certain that `G._adj[u].keys()` returns only unique neighbors (even in a `MultiGraph`), ignoring multiple edges, and also, as you have mentioned, the intersection is a set of nodes and not edges. For that reason, I just documented the behavior for multigraphs, and have also updated the tests accordingly!

**Review by rossbar — APPROVED** [review]
I took the liberty of pushing up a few minor changes related to formatting etc. but this LGTM! Thanks @georakom and @amcandio + @Peiffap for the detailed review!
